### ISSUE TITLE
Cookies not being cleared on quit

### ISSUE THREAD

**Issue by tcz305** [bromite/bromite#142]
**Bromite version**

Version: 69.0.3497.109
Arch: arm64
Android version: Oreo

**Is the bug reproducible with latest version?**
Yes

**Can the bug be reproduced with corresponding [Chromium](https://github.com/bromite/chromium/releases) version?**
No

**Is the bug a crash?**
No

**Describe the bug**
With the setting "keep cookies only until you quit your browser" checked, and logging into a site eg amazon etc, and later exiting the browser by clicking exit, login remains through restarts **unless site storage is cleared**.
All sites also remember all info between sessions like store location entered, site browsing history, etc.
_I'm not sure if this is by design that all info is being stored in site storage but I thought clearing cookies would clear all the info_.

**To Reproduce**
1. Go to any site
2. Click on login or enter any other info
3. Exit browser
4. On restart appears all cookies (seem to) remain.

**Expected behavior**
Expected all cookies to be cleared including logins and entered info and site browsing history and location etc.

**Comment by Techguyprivate** [comment]
Where is the setting for session cookies? Can you post a screenshot?

**Comment by csagan5** [comment]
@tcz305 a session is closed when the app exits. You need to forcefully stop it to be sure the session ended.

**Comment by csagan5** [comment]
`69.0.3497.91` also affected.

**Comment by csagan5** [comment]
Is this problem still reproducible?

**Comment by Adam-Stambaugh** [comment]
I have the same issue on my end. Cookies aren't clearing when I quit out of the browser with the option enabled as shown below.
![Screenshot_20190920-133357_Bromite](https://user-images.githubusercontent.com/49969486/65350660-d15e0080-dbab-11e9-97c2-7d2eeb72cfaa.png)
I'm running LineageOS 16.0, WebView (AOSP) 76.0.3809.111, and Bromite 76.0.3865.81.

**Comment by sviluppoSoftware** [comment]
i confirm this bug it's still present in 78.0.3904.130 version too

**Comment by kef71** [comment]
I can confirm this bug it's still present in v79.0.3945.123

**Comment by uazo** [comment]
It can be done at startup, after `ChromeActivity.finishNativeInitialization()` adding for example

```
BrowsingDataBridge.getInstance().clearBrowsingData(
        new BrowsingDataBridge.OnClearBrowsingDataListener() {
	   @Override
            public void onBrowsingDataCleared() {
		Context context = ContextUtils.getApplicationContext();
		Toast toast = Toast.makeText(context, "Data cleanup", Toast.LENGTH_SHORT);
		toast.show();
            }
        },
	new int[] {
		BrowsingDataType.CACHE
		,BrowsingDataType.COOKIES
		,BrowsingDataType.FORM_DATA
		//,BrowsingDataType.HISTORY
		//,BrowsingDataType.PASSWORDS
		//,BrowsingDataType.SITE_SETTINGS
	}, TimePeriod.ALL_TIME);
```

**Comment by csagan5** [comment]
@uazo the patch that implements this feature (in its current state) is this: https://github.com/bromite/bromite/blob/81.0.4044.106/build/patches/prefs-only-keep-cookies-until-exit.patch

Unfortunately `CONTENT_SETTING_SESSION_ONLY` is not effective on Android, while it seems to be effective on desktop versions which use this same setting (from my analysis back then).

Rather than adding code we should fix the behaviour of `CONTENT_SETTING_SESSION_ONLY`? The fix you propose is probably more in line with #426 which we would not need if this was fixed.

**Comment by uazo** [comment]
> Rather than adding code we should fix the behaviour of CONTENT_SETTING_SESSION_ONLY? 

I try, but unfortunately I have to set up a serious development machine first. is there any CI on gitlab?

> The fix you propose is probably more in line with #426 which we would not need if this was fixed.

do you mean here?
https://github.com/bromite/bromite/blob/master/build/patches/Add-exit-menu-item.patch

**Comment by Pavbc** [comment]
> which we would not need if this was fixed.

If i may, my understanding as a user is that the options described in [#426](https://github.com/bromite/bromite/issues/426) by codeboy66 and in [#530](https://github.com/bromite/bromite/issues/530) by myself are more useful that just clear cookies on exit. 

Although both solution may be similar and the "clear cookies" is probably easy to implement, clearing ALL data (including history, cache, open tabs, etc and not just cookies ***) only when the user presses the exit button seem to be a much more useful tool. For instance, it allows for a better continuity between browsing sessions without the Hassel of being unlogged just to login again 10 minutes later and still having the option to, when finished, clear all data quickly.

***Both sugentions #426 and #530 say that the "data" cleared could be the one checked on history menu

Nevertheless, if the developer choose not to implement this, it's ok as well, the always incognito is really fine and still the fastest browser I've used. Thanks for your time

**Comment by csagan5** [comment]
> is there any CI on gitlab?

No, but there have been attempts [here](https://github.com/Eloston/ungoogled-chromium/issues/916) and [here](https://github.com/wchen342/ungoogled-chromium-android/issues/8#issuecomment-616134415)

> do you mean here?

let's discuss it on issue #530

**Comment by wchen342** [comment]
> > is there any CI on gitlab?
> 
> No, but there have been attempts [here](https://github.com/Eloston/ungoogled-chromium/issues/916) and [here](https://github.com/Eloston/ungoogled-chromium/issues/845)

The process is currently blocked because of an unfinished functionality in Github Actions. But the .yml file can be used as a reference for setting up a docker machine. Although I don't think using CI for developing chromium is a good idea considering the compiling time.

**Comment by uazo** [comment]
> Unfortunately CONTENT_SETTING_SESSION_ONLY is not effective on Android, while it seems to be effective on desktop versions which use this same setting (from my analysis back then).

Thanks for your patience .. So, I found out that

- the management of session cookies relative to the patch you have indicated is incorrect, simply because CONTENT_SETTING_SESSION_ONLY is never set
- the creation of the session cookies deletion function (session_cleanup_cookie_store) is done during the creation of the cookie_store (SQLitePersistentCookieStore) in MakeURLRequestContext() but
- the DeleteSessionCookies() of session_cleanup_cookie_store is called to the cookie_manager destructor 

the problem is: cookie manager dtor() is never invoked, so that call is never made.
There is something about application shutdown here https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/lifetime/application_lifetime.cc but the jni bridge for java doesn't exist unless you implement it (I'll try because the patch call https://github.com/bromite/bromite/blob/master/build/patches/Add-exit-menu-item.patch#L42 is conceptually incorrect, that method seems to be called by the native and not by java).

I'm about to send a pull request to correctly manage session cookies (compiling time!), The changes are a bit many, however, considering that from the M86 the flag becomes effective (https://chromium.googlesource.com/chromium/src/+/8182dea5f76ab41064cb968ba06409bc3e771dd5) I think there is no other solution, however:

1. Invoking that method could be an acceptable solution but in android applications it does not seem possible to manage the user shutdown event from the recents, because zygote kills the task with sigkill which is not trappable, so that function must be call at initialization time. Same thing happens with windows version, if you kill the task cookies are not deleted. I have seen android solutions using services, but I think they are not applicable because I imagine that the native is no longer accessible.
2. What is deleted are only cookies, of course, and personally I find it of little use, considering that all other data are kept.

If you agree, developing this changes, I have seen that it is possible to recreate what is done here https://github.com/Cookie-AutoDelete/Cookie-AutoDelete with a mix of changes between java and c++. Could it be an acceptable solution?

**Comment by csagan5** [comment]
> * the management of session cookies relative to the patch you have indicated is incorrect, simply because CONTENT_SETTING_SESSION_ONLY is never set

That patch only changes a default, I do not know if it has any effect right now; `CONTENT_SETTING_SESSION_ONLY` is set through another patch, see https://github.com/bromite/bromite/blob/84.0.4147.119/build/patches/Add-site-settings-option-for-session-only-cookies.patch

> the patch call https://github.com/bromite/bromite/blob/master/build/patches/Add-exit-menu-item.patch#L42 is conceptually incorrect, that method seems to be called by the native and not by java

Is this something that needs another fix?

> considering that from the M86 the flag becomes effective (https://chromium.googlesource.com/chromium/src/+/8182dea5f76ab41064cb968ba06409bc3e771dd5)

Is upstream going to make changes that render your patch obsolete/conflicting?

> 1. Invoking that method could be an acceptable solution

It is acceptable to perform the action on next startup on Android, as far as we document this behaviour.

> 2\. What is deleted are only cookies, of course, and personally I find it of little use, considering that all other data are kept.

If we were to delete other data it would be more akin to #530; the idea of `CONTENT_SETTING_SESSION_ONLY` is that you can even set it for a specific site, so when a new session is started only the cookies of sites which have `CONTENT_SETTING_SESSION_ONLY` would get deleted (or all, if `CONTENT_SETTING_SESSION_ONLY` is set globally).



> If you agree, developing this changes, I have seen that it is possible to recreate what is done here https://github.com/Cookie-AutoDelete/Cookie-AutoDelete with a mix of changes between java and c++. Could it be an acceptable solution?

Would this approach respect the `CONTENT_SETTING_SESSION_ONLY` setting for each site? Have you looked at the other patch https://github.com/bromite/bromite/blob/84.0.4147.119/build/patches/Add-site-settings-option-for-session-only-cookies.patch ?

**Comment by uazo** [comment]
> That patch only changes a default, I do not know if it has any effect right now; CONTENT_SETTING_SESSION_ONLY is set through another patch, see https://github.com/bromite/bromite/blob/84.0.4147.119/build/patches/Add-site-settings-option-for-session-only-cookies.patch

the bug is that internally the value is reset after calling https://github.com/bromite/bromite/blob/96671ab221f05b5615dd897e8580c31e375c08da/build/patches/Add-site-settings-option-for-session-only-cookies.patch#L59

so that patch is useless

> Is this something that needs another fix?

yes and no, it should be checked.
yes because it is marked as called by the native, which means that there is something else that needs to be launched, so that the native resources are first freed and then the java part closed.
no because it seems to be irrelevant and not currently managed even upstream

> Is upstream going to make changes that render your patch obsolete/conflicting?

No, that patch is ready for _current_ master.
Only this file /components/content_settings/core/common/features.cc must be changed:
https://github.com/bromite/bromite/pull/671/files#diff-cd6113d00a34caf5f81191e4f8559385R441

> It is acceptable to perform the action on next startup on Android, as far as we document this behaviour.

removal is effective when Bromite is closed with its Exit button. I added cleanup to the first request as well, in order to handle closing from recents, effectively messing up the code a bit .. Nothing scary, but technically mojo changes need an ok from team security, but the change is not impacting.

> is that you can even set it for a specific site, so when a new session is started only the cookies of sites which have CONTENT_SETTING_SESSION_ONLY would get deleted (or all, if CONTENT_SETTING_SESSION_ONLY is set globally).

er, no, at least not currently.
the android ui does not allow the application of session cookies as a per site restriction, but it is something that can be done.

> Would this approach respect the CONTENT_SETTING_SESSION_ONLY setting for each site? 

the idea is to add the possibility of automatic deletion of cookies, local storage, indexed db and everything currently managed from here https://source.chromium.org/chromium/chromium/src/+/master:content/browser/browsing_data/browsing_data_remover_impl.cc;l=259 with appropriate flags, when the tab is closed, for all the browsing performed from that tab.
Of course it is possible to think of whitelists or something similar

The changes would certainly impact this
https://source.chromium.org/chromium/chromium/src/+/master:chrome/android/java/src/org/chromium/chrome/browser/tab/TabWebContentsObserver.java and saving and loading of tabs from the profile user.

**Comment by csagan5** [comment]
> the bug is that internally the value is reset after calling
> 
> https://github.com/bromite/bromite/blob/96671ab221f05b5615dd897e8580c31e375c08da/build/patches/Add-site-settings-option-for-session-only-cookies.patch#L59
> 
> so that patch is useless

The other patch I mentioned also changes how the session only value is set.

> > Is upstream going to make changes that render your patch obsolete/conflicting?

> No, that patch is ready for _current_ master.
> Only this file /components/content_settings/core/common/features.cc must be changed:

There is some misunderstanding; you mentioned that upstream is improving the cookie controls, so I wonder if there will be similar functionality to your PR but developed by upstream?

Also what do you mean "only this file" must be changed? What about the rest of the changes in the patch?

>>  the idea of `CONTENT_SETTING_SESSION_ONLY` is that you can even set it for a specific site, so when a new session is started only the cookies of sites which have CONTENT_SETTING_SESSION_ONLY would get deleted (or all, if CONTENT_SETTING_SESSION_ONLY is set globally).

> er, no, at least not currently.
> the android ui does not allow the application of session cookies as a per site restriction, but it is something that can be done.

That is already implemented in the patch I mentioned in previous comment: https://github.com/bromite/bromite/blob/84.0.4147.119/build/patches/Add-site-settings-option-for-session-only-cookies.patch 

> removal is effective when Bromite is closed with its Exit button. I added cleanup to the first request as well, in order to handle closing from recents, effectively messing up the code a bit

Then we need to document that sometimes cookies are cleared on exit, sometimes on next startup.

>> Would this approach respect the CONTENT_SETTING_SESSION_ONLY setting for each site?
> the idea is to add the possibility of automatic deletion of cookies, local storage, indexed db and everything currently managed from here

So the answer is no, the per-site setting would not be respected.
For the purpose you mentioned there is already incognito mode, I am not sure if it's viable having incognito mode and the regular mode with flags to forget all those.

I assume your PR is built on top of the patch https://github.com/bromite/bromite/blob/84.0.4147.119/build/patches/Add-site-settings-option-for-session-only-cookies.patch 

Can you please check whether those changes are still necessary with your changes?

Thanks again for your work!

**Comment by uazo** [comment]
> The other patch I mentioned also changes how the session only value is set.

Do you mean this https://github.com/bromite/bromite/blob/master/build/patches/prefs-only-keep-cookies-until-exit.patch ?
Unfortunately, that too is not good. The default should not be set to session_only simply because the value is reset internally to allow.
And since allow is different from session_only, it leads to the generation of a sort of navigation tracking in Preferences.json (in the exceptions section of the content_settings) when the flag is unchecked and rechecked, at least until the restart.

> There is some misunderstanding; you mentioned that upstream is improving the cookie controls, so I wonder if there will be similar functionality to your PR but developed by upstream?

sorry I did not understand
no, upstream will enable advanced management of cookies, but there is no possibility to set session_only.

> Also what do you mean "only this file" must be changed? What about the rest of the changes in the patch?

I mean that when you get to that version, my git-diff will no longer be fine, it will have to be checked for that file.
the rest of the changes are still fine for now, I can't say in the future.
(OT by the way, out of curiosity, but do you fix any problems with git-apply by hand or do you use any particular program?)

> That is already implemented in the patch I mentioned in previous comment:

that patch doesn't work.
I updated the commit with the ability to define it per-site

> For the purpose you mentioned there is already incognito mode, I am not sure if it's viable having incognito mode and the regular mode with flags to forget all those.

always incognito does not record history, and will never record it because the user profile is not persistent by design (OffTheRecord in code).
my proposal would allow an incognito navigation but with the history active and cleaning simply by closing the tabs, among other things without the annoyance of that permanent notification that warns me if I want to close all incognito tabs.

> Can you please check whether those changes are still necessary with your changes?

no, that code is no longer called by the rest of the application, if you want you can remove everything, or leave it, nothing changes

**Comment by csagan5** [comment]
> no, upstream will enable advanced management of cookies, but there is no possibility to set session_only.

I have already enabled the improved cookie controls in `84.0.4147.121`.

> always incognito does not record history, and will never record it because the user profile is not persistent by design (OffTheRecord in code).
my proposal would allow an incognito navigation but with the history active and cleaning simply by closing the tabs, among other things without the annoyance of that permanent notification that warns me if I want to close all incognito tabs.

> cleaning simply by closing the tabs

You mean that cleaning happens when the individual tab is closed, your patch already works this way?

I am now building Bromite with your patch; regarding:

> the idea is to add the possibility of automatic deletion of cookies, local storage, indexed db and everything currently managed from here https://source.chromium.org/chromium/chromium/src/+/master:content/browser/browsing_data/browsing_data_remover_impl.cc;l=259 with appropriate flags, when the tab is closed, for all the browsing performed from that tab.

More flags would be needed. Perhaps a single new flag which covers everything else except history? I do not currently know if history can be tracked on Javascript-side (CSS `:visited:` and such), but I guess we can leave that out for now.

Let me know what is your plan for this.

**Comment by uazo** [comment]
> You mean that cleaning happens when the individual tab is closed, your patch already works this way?

No, developing that is a bit complex, it takes time.

> Let me know what is your plan for this.

I made a mistake, enabling history for OTR profiles using a flag is pretty trivial. I will open an issue and a pr

**Comment by csagan5** [comment]
> I made a mistake, enabling history for OTR profiles using a flag is pretty trivial. I will open an issue and a pr

Interesting, looking forward to see that.

**Comment by uazo** [comment]
fixed with https://github.com/bromite/bromite/pull/671
or not?

**Comment by csagan5** [comment]
Yes; the best would be to close these issues automatically with the release commit, or we close them by writing `fixes #...` in the commit contained in the PR.

Fixed in `84.0.4147.132`.
